src/cmfd_solver.cpp

```diff
@@ -0,0 +1,361 @@
+#include "openmc/cmfd_solver.h"
+
+#include <vector>
+#include <cmath>
+
+#include "xtensor/xtensor.hpp"
+
+#include "openmc/error.h"
+#include "openmc/constants.h"
+#include "openmc/capi.h"
+
+namespace openmc {
+
+namespace cmfd {
+
+//==============================================================================
+// Global variables
+//==============================================================================
+
+std::vector<int> indptr;
+
+std::vector<int> indices;
+
+int dim;
+
+double spectral;
+
+int nx, ny, nz, ng;
+
+xt::xtensor<int, 2> indexmap;
+
+} // namespace cmfd
+
+//==============================================================================
+// MATRIX_TO_INDICES converts a matrix index to spatial and group
+// indices
+//==============================================================================
+
+void matrix_to_indices(int irow, int& g, int& i, int& j, int& k)
+{
+  g = irow % cmfd::ng;
+  i = cmfd::indexmap(irow/cmfd::ng, 0);
+  j = cmfd::indexmap(irow/cmfd::ng, 1);
+  k = cmfd::indexmap(irow/cmfd::ng, 2);
+}
+
+//==============================================================================
+// GET_DIAGONAL_INDEX returns the index in CSR index array corresponding to
+// the diagonal element of a specified row
+//==============================================================================
+
+int get_diagonal_index(int row)
+{
+  for (int j = cmfd::indptr[row]; j < cmfd::indptr[row+1]; j++) {
+    if (cmfd::indices[j] == row)
+      return j;
+  }
+
+  // Return -1 if not found
+  return -1;
+}
+
+//==============================================================================
+// SET_INDEXMAP sets the elements of indexmap based on input coremap
+//==============================================================================
+
+void set_indexmap(const int* coremap)
+{
+  for (int z = 0; z < cmfd::nz; z++) {
+    for (int y = 0; y < cmfd::ny; y++) {
+      for (int x = 0; x < cmfd::nx; x++) {
+        if (coremap[(z*cmfd::ny*cmfd::nx) + (y*cmfd::nx) + x] != CMFD_NOACCEL) {
```

Comment:
This code looks correct, though these two lines could be a little more readable if the index were precomputed, eg:
```c++
        int idx = (z*cmfd::ny*cmfd::nx) + (y*cmfd::nx) + x;
        if (coremap[idx] != CMFD_NOACCEL) {
          int counter = coremap[idx];
```